app/assets/javascripts/discourse/app/components/d-button.js

```diff
@@ -1,161 +1,119 @@
+  get btnType() {
+    if (this.args.icon) {
+      return this.computedLabel ? "btn-icon-text" : "btn-icon";
+    } else if (this.computedLabel) {
       return "btn-text";
     }
-  },
+  }
 
-  noText: empty("computedLabel"),
-
-  @discourseComputed("title", "translatedTitle")
-  computedTitle(title, translatedTitle) {
-    if (title) {
-      return I18n.t(title);
+  get computedTitle() {
+    if (this.args.title) {
+      return I18n.t(this.args.title);
     }
-    return translatedTitle;
-  },
+    return this.args.translatedTitle;
+  }
 
-  @discourseComputed("label", "translatedLabel")
-  computedLabel(label, translatedLabel) {
-    if (label) {
-      return I18n.t(label);
+  get computedLabel() {
+    if (this.args.label) {
+      return I18n.t(this.args.label);
     }
-    return translatedLabel;
-  },
+    return this.args.translatedLabel;
+  }
 
-  @discourseComputed("ariaLabel", "translatedAriaLabel")
-  computedAriaLabel(ariaLabel, translatedAriaLabel) {
-    if (ariaLabel) {
-      return I18n.t(ariaLabel);
+  get computedAriaLabel() {
+    if (this.args.ariaLabel) {
+      return I18n.t(this.args.ariaLabel);
     }
-    if (translatedAriaLabel) {
-      return translatedAriaLabel;
+    if (this.args.translatedAriaLabel) {
+      return this.args.translatedAriaLabel;
     }
-  },
+  }
 
-  @discourseComputed("ariaExpanded")
-  computedAriaExpanded(ariaExpanded) {
-    if (ariaExpanded === true) {
+  get computedAriaExpanded() {
+    if (this.args.ariaExpanded === true) {
       return "true";
     }
-    if (ariaExpanded === false) {
+    if (this.args.ariaExpanded === false) {
       return "false";
     }
-  },
+  }
 
+  @action
   keyDown(e) {
-    if (this.onKeyDown) {
+    if (this.args.onKeyDown) {
       e.stopPropagation();
-      this.onKeyDown(e);
+      this.args.onKeyDown(e);
     } else if (e.key === "Enter") {
       this._triggerAction(e);
-      return false;
     }
-  },
+  }
 
+  @action
   click(event) {
     return this._triggerAction(event);
-  },
+  }
 
+  @action
   mouseDown(event) {
-    if (this.preventFocus) {
+    if (this.args.preventFocus) {
       event.preventDefault();
     }
-  },
+  }
 
   _triggerAction(event) {
-    let { action, route, href } = this;
+    const { action: actionVal, route, href } = this.args;
 
-    if (action || route || href?.length) {
-      if (action) {
-        if (typeof action === "string") {
+    if (actionVal || route || href?.length) {
+      if (actionVal) {
+        const { actionParam, forwardEvent } = this.args;
+
+        if (typeof actionVal === "string") {
           // Note: This is deprecated in new Embers and needs to be removed in the future.
           // There is already a warning in the console.
-          this.sendAction("action", this.actionParam);
-        } else if (typeof action === "object" && action.value) {
-          if (this.forwardEvent) {
-            action.value(this.actionParam, event);
+          this.sendAction("action", actionParam);
```

Comment:
`sendAction` is not a function on Glimmer components, so this throws an error. Can you think of any way to make this work @dgeb ?

I realise `sendAction` is deprecated and will be removed in Ember 4, so maybe we just need to bite the bullet here?

The case we noticed is

https://github.com/discourse/discourse/blob/7e992cb29919c272b202ec599da0014eb834954b/app/assets/javascripts/discourse/app/templates/bulk-tag.hbs#L5-L5

(`this.action` is a string)

I'll look into modernising that case now. But there might be others, especially in themes/plugins.

Reply:
@davidtaylorhq For better or worse, I don't think that a clean workaround is possible for glimmer components without requiring that a target context be passed to the component. And if modification is needed at the usage site, then it would be far preferable to refactor to a closure action (as you did for `bulk-tag`) rather than to prop up the old usage pattern.

Obviously, we should no longer be calling `this.sendAction` from within a glimmer component. In the case that `action` is passed as a string, would you like me to call `deprecated` (where `deprecated` comes from "discourse-common/lib/deprecated";`)? Or perhaps raising an error would be more appropriate?

Reply:
Thanks @dgeb - that makes sense. Could we call `deprecated` on initial render, and then throw an exception when the button is clicked? That way it's nice and clear without totally breaking rendering.

Reply:
@davidtaylorhq sure, that sounds like a solid approach 👍

Reply:
@davidtaylorhq ok, this PR has been updated.